tools/chainsync-loadtest/src/concurrency/scope.rs

```diff
@@ -0,0 +1,184 @@
+    err: Option<anyhow::Error>,
+    cancelled: bool,
+}
+
+// Scope represents a collection of spawned futures with a common bounded lifetime.
+// See Scope::run for details.
+pub struct Scope {
+    ctx: CtxWithCancel,
+    // Main futures have to exit before the scope is cancelled.
+    main_futures: WaitGroup,
+    // Additional futures running in the background,
+    // which will be cancelled as soon as main_futures exit.
+    weak_futures: WaitGroup,
+    state: Mutex<ScopeState>,
+}
+
+impl Scope {
+    fn complete(&self, v: anyhow::Result<()>) {
+        let e = if let Err(e) = v {
+            e
+        } else {
+            return;
+        };
+        let mut s = self.state.lock().unwrap();
+        if s.err.is_some() {
+            return;
+        }
+        // Ignore Cancelled errors after the whole scope has been cancelled.
+        if s.cancelled
+            && e.downcast_ref::<ctx::Error>().map(|e| e == &ctx::Error::Cancelled).unwrap_or(false)
+        {
+            return;
+        }
+        s.err = Some(e);
+        s.cancelled = true;
+        self.ctx.cancel();
+    }
+
+    // Spawn a "main" future in the scope.
+    pub fn spawn<F>(self: &Arc<Self>, f: impl Send + FnOnce(Ctx, Arc<Self>) -> F)
+    where
+        F: Future<Output = anyhow::Result<()>> + Send + 'static,
+    {
+        let fut = f((*self.ctx).clone(), self.clone());
+        let s = self.clone();
+        s.main_futures.inc();
+        tokio::spawn(async move {
+            s.complete(fut.await);
+            s.main_futures.dec();
+        });
+    }
+
+    // Spawn a "weak" future in the scope.
+    pub fn spawn_weak<F>(self: &Arc<Self>, f: impl Send + FnOnce(Ctx) -> F)
+    where
+        F: Future<Output = anyhow::Result<()>> + Send + 'static,
+    {
+        let fut = f((*self.ctx).clone());
+        let s = self.clone();
+        s.weak_futures.inc();
+        tokio::spawn(async move {
+            s.complete(fut.await);
+            s.weak_futures.dec();
+        });
+    }
+
+    // Scope represents a collection of futures executed concurrently on a tokio runtime, which share a context.
+    // Scope allows for cooperative execution, therefore all coroutines are allowed to run as long
+    // as they want. If you cancel the context, the futures will get notified (see documentation of
+    // Ctx) but won't get preempted: they will be able to shutdown gracefully.
+    //
+    // run() is a lambda-abstraction of the Scope:
+    // 1. it creates a new scope, spawns f(ctx,scope) and adds it to the scope.
+    // 2. This first future of the scope is allowed to spawn additional futures in the scope.
+    // 3. run() is blocking until all members of the scope complete.
+    //
+    // Since from the point of view of the caller of Scope::run, it is just a blocking call, which
+    // might spawn coroutines internally, but all of them will complete before Scope::run returns,
+    // therefore Scope::run allows for expressing STRUCTURED CONCURRENCY.
```

Comment:
**Important Note** this is going to be only best-effort structured concurrency, due to how cancellation-via-drop works, it's not really possible to express water-tight structured concurrency in `async` Rust, as far as I am aware. Here's a test that shows a task outliving it's scope:

```rust
#[tokio::test]
async fn break_structured_concurrency() {
    {
        let ctx = Ctx::background();
        let scope = Scope::run(&ctx, |_ctx, scope| async move {
            scope.spawn(|_ctx, scope| async move {
                loop {
                    tokio::time::sleep(Duration::from_millis(500)).await;
                    eprintln!("still running");
                }
            });
            Ok(())
        });
        let sleep = tokio::time::sleep(Duration::from_millis(100));
        tokio::pin!(sleep);
        tokio::select! {
            biased;
            _ = scope => {}
            _ = sleep => {}
        }

    }

    loop {
        tokio::task::yield_now().await
    }
}
```